File: tests/test_omnipress.js
/**
 * Comprehensive Unit Tests for OmniPress MCP (Node.js engine)
 */

const assert = require('node:assert');
const fs = require('node:fs');
const path = require('node:path');
const os = require('node:os');
const {
  inspectSensitiveData,
  sanitizeText,
  slugify,
  saveArticle,
  handleRpcRequest,
  TOOLS_DEFINITIONS,
  createHttpServer
} = require('../src/server.js');

async function runTests() {
  console.log('🧪 Starting OmniPress MCP Unit Test Suite...\n');

  // Test 1: Sanitizer & Legal / Tax PII Detection
  console.log('Test 1: Sanitizer & PII Detection...');
  const sampleText =
    'No processo 0001234-56.2023.8.26.0100 o perito analisou a empresa CNPJ 12.345.678/0001-90 ' +
    'onde o sócio CPF 123.456.789-00 respondeu via perito@tribunal.jus.br sobre a perícia.';

  const findings = inspectSensitiveData(sampleText);
  assert.strictEqual(findings.length, 4, `Expected 4 sensitive items, found ${findings.length}`);
  const foundTypes = findings.map(f => f.type);
  assert(foundTypes.includes('lawsuit_number'), 'Must detect lawsuit number');
  assert(foundTypes.includes('cnpj'), 'Must detect CNPJ');
  assert(foundTypes.includes('cpf'), 'Must detect CPF');
  assert(foundTypes.includes('email'), 'Must detect email');

  const { sanitized } = sanitizeText(sampleText);
  assert(!sanitized.includes('0001234-56.2023.8.26.0100'), 'Lawsuit number should be masked');
  assert(!sanitized.includes('12.345.678/0001-90'), 'CNPJ should be masked');
  assert(!sanitized.includes('123.456.789-00'), 'CPF should be masked');
  assert(!sanitized.includes('perito@tribunal.jus.br'), 'Email should be masked');
  assert(sanitized.includes('[PROCESSO ANÔNIMO]'), 'Expected [PROCESSO ANÔNIMO] mask');
  console.log('  ✓ Sanitizer passed: All CNJ lawsuit numbers, CPFs, CNPJs, and emails masked.\n');

  // Test 2: Slugify & URL Normalization
  console.log('Test 2: Slugify...');
  assert.strictEqual(slugify('Reforma Tributária e Software: O que Muda?'), 'reforma-tributaria-e-software-o-que-muda');
  assert.strictEqual(slugify('Auditando Bancos de Dados em Perícia Judicial'), 'auditando-bancos-de-dados-em-pericia-judicial');
  console.log('  ✓ Slugify passed.\n');

  // Test 3: Article Archiving with YAML Frontmatter
  console.log('Test 3: Article Archiving...');
  const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'omnipress-test-'));
  process.env.ARTICLES_DIR = tempDir;

  const saveReport = saveArticle({
    title: 'Auditando Bancos de Dados em Perícia Judicial',
    contentMarkdown: 'Conteúdo detalhado da perícia contábil e de TI.',
    category: 'pericia-judicial',
    tags: ['pericia', 'sql', 'tributario'],
    author: 'Perito Judicial TI'
  });

  assert(saveReport.includes('Article archived successfully'), 'Expected success message');
  const files = fs.readdirSync(tempDir, { recursive: true });
  const mdFiles = files.filter(f => f.toString().endsWith('.md'));
  assert(mdFiles.length > 0, 'Markdown article should have been created on disk');

  const savedContent = fs.readFileSync(path.join(tempDir, mdFiles[0]), 'utf8');
  assert(savedContent.includes('title: "Auditando Bancos de Dados em Perícia Judicial"'));
  assert(savedContent.includes('category: "pericia-judicial"'));
  assert(savedContent.includes('- pericia'));
  assert(savedContent.includes('status: draft'));
  console.log('  ✓ Article archiving passed: File created with valid YAML frontmatter.\n');

  // Test 4: JSON-RPC MCP Protocol Handshake & Tool Listing
  console.log('Test 4: MCP Protocol Handshake...');
  const initRes = await handleRpcRequest({
    jsonrpc: '2.0',
    id: 1,
    method: 'initialize',
    params: { capabilities: {} }
  });
  assert.strictEqual(initRes.result.serverInfo.name, 'omnipress');
  assert.strictEqual(initRes.result.protocolVersion, '2024-11-05');

  const toolsRes = await handleRpcRequest({
    jsonrpc: '2.0',
    id: 2,
    method: 'tools/list',
    params: {}
  });
  assert.strictEqual(toolsRes.result.tools.length, 4);
  const toolNames = toolsRes.result.tools.map(t => t.name);
  assert(toolNames.includes('omnipress_publish_article'));
  assert(toolNames.includes('omnipress_queue_post'));
  assert(toolNames.includes('omnipress_inspect_content'));
  assert(toolNames.includes('omnipress_list_recent'));
  console.log('  ✓ MCP protocol passed: initialize and tools/list resolved correctly.\n');

  // Test 5: Tool Call Execution via JSON-RPC
  console.log('Test 5: MCP Tool Call Execution (omnipress_inspect_content)...');
  const callRes = await handleRpcRequest({
    jsonrpc: '2.0',
    id: 3,
    method: 'tools/call',
    params: {
      name: 'omnipress_inspect_content',
      arguments: {
        text: 'Documento limpo sem nenhum dado confidencial.'
      }
    }
  });
  assert(callRes.result.content[0].text.includes('Content is clean!'));
  console.log('  ✓ MCP tool call passed.\n');

  // Test 6: Threads & Reddit Tool Call Execution via JSON-RPC
  console.log('Test 6: MCP Queue Post (Threads & Reddit)...');
  const queueRes = await handleRpcRequest({
    jsonrpc: '2.0',
    id: 4,
    method: 'tools/call',
    params: {
      name: 'omnipress_queue_post',
      arguments: {
        text: 'Lançando nova ferramenta aberta para perícia judicial de software no processo 0001234-56.2023.8.26.0100!',
        platforms: ['threads', 'reddit:SideProject'],
        title: 'Open Source Software Forensics Tool for Antigravity'
      }
    }
  });
  const queueOutput = queueRes.result.content[0].text;
  assert(queueOutput.includes('[PROCESSO ANÔNIMO]'), 'Sensitive court case number must be masked');
  assert(queueOutput.includes('Masked 1 sensitive item'), 'Should report 1 masked item');
  console.log('  ✓ Threads & Reddit queue test passed: Sanitization & parameters validated.\n');

  // Clean up test directory
  fs.rmSync(tempDir, { recursive: true, force: true });

  console.log('═══════════════════════════════════════════════════════');
  console.log('🎉 ALL 6 TEST SUITES PASSED WITH MAXIMUM RIGOR & CARE!');
  console.log('═══════════════════════════════════════════════════════');
}

runTests().catch(err => {
  console.error('❌ Test suite failed:', err);
  process.exit(1);
});

File: tests/test_omnipress.py
"""Unit tests for OmniPress MCP core modules."""

import sys
from pathlib import Path

# Add project root to sys.path
PROJECT_ROOT = Path(__file__).resolve().parent.parent
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

from src.utils.sanitizer import sanitize_text, inspect_sensitive_data
from src.adapters.markdown_adapter import MarkdownArchiveAdapter, slugify


def test_sanitizer_detection():
    sample_text = (
        "No processo 0001234-56.2023.8.26.0100 o perito analisou o CNPJ 12.345.678/0001-90 "
        "e o autor com CPF 123.456.789-00 enviou email para perito@tribunal.jus.br."
    )
    findings = inspect_sensitive_data(sample_text)
    assert len(findings) == 4, f"Expected 4 sensitive items, got {len(findings)}"

    sanitized, _ = sanitize_text(sample_text)
    assert "0001234-56.2023.8.26.0100" not in sanitized
    assert "12.345.678/0001-90" not in sanitized
    assert "123.456.789-00" not in sanitized
    assert "perito@tribunal.jus.br" not in sanitized
    assert "[PROCESSO ANÔNIMO]" in sanitized
    print("✓ Sanitizer test passed successfully.")


def test_slugify():
    assert slugify("Reforma Tributária e Software: O que Muda?") == "reforma-tributaria-e-software-o-que-muda"
    assert slugify("Test 123! & Special Chars") == "test-123-special-chars"
    print("✓ Slugify test passed successfully.")


def test_markdown_adapter(tmp_path: Path):
    adapter = MarkdownArchiveAdapter(base_dir=str(tmp_path))
    res = adapter.save_article(
        title="Auditando Bancos de Dados em Perícia Judicial",
        content_markdown="Aqui está o conteúdo técnico da perícia...",
        category="pericia-judicial",
        tags=["pericia", "sql", "forense"],
    )

    assert res["success"] is True
    saved_file = Path(res["filepath"])
    assert saved_file.exists()

    content = saved_file.read_text(encoding="utf-8")
    assert 'title: "Auditando Bancos de Dados em Perícia Judicial"' in content
    assert "category: \"pericia-judicial\"" in content
    assert "- pericia" in content

    recent = adapter.list_recent(limit=5)
    assert len(recent) == 1
    assert recent[0]["name"] == res["filename"]
    print("✓ Markdown adapter test passed successfully.")


def test_mcp_server():
    from src.server import (
        MCP_TOOL_ANNOTATIONS,
        omnipress_publish_article,
        omnipress_queue_post,
        omnipress_inspect_content,
        omnipress_list_articles,
        omnipress_list_recent,
    )

    required_tools = [
        "omnipress_publish_article",
        "omnipress_queue_post",
        "omnipress_inspect_content",
        "omnipress_list_articles",
        "omnipress_list_recent",
    ]

    for tool_name in required_tools:
        assert tool_name in MCP_TOOL_ANNOTATIONS, f"Missing annotation for {tool_name}"
        annot = MCP_TOOL_ANNOTATIONS[tool_name]
        for key in ("readOnlyHint", "destructiveHint", "idempotentHint", "openWorldHint"):
            assert isinstance(annot.get(key), bool), f"{key} in {tool_name} must be a boolean"

    # Verify structured docstrings
    for func in (omnipress_publish_article, omnipress_queue_post, omnipress_inspect_content, omnipress_list_articles):
        doc = func.__doc__ or ""
        assert "Use when:" in doc, f"{func.__name__} docstring missing 'Use when:'"
        assert "Do NOT use when:" in doc, f"{func.__name__} docstring missing 'Do NOT use when:'"
        assert "Returns:" in doc, f"{func.__name__} docstring missing 'Returns:'"

    # Verify alias parity
    canonical_res = omnipress_list_articles(limit=5)
    alias_res = omnipress_list_recent(limit=5)
    assert canonical_res == alias_res, "omnipress_list_articles and omnipress_list_recent must yield identical output"
    print("✓ MCP server annotations, docstrings, and alias parity passed successfully.")


if __name__ == "__main__":
    test_sanitizer_detection()
    test_slugify()
    import tempfile
    with tempfile.TemporaryDirectory() as tmpdir:
        test_markdown_adapter(Path(tmpdir))
    test_mcp_server()
    print("\n🎉 ALL TESTS PASSED SUCCESSFULLY!")
